vulnerabilities/overflow-underflow.md

```diff
@@ -1,73 +1,158 @@
 
 ### Using Shift Operators
-Overflow & underflow checks are not performed for shift operations like they are performed for other arithmetic operations. Thus, over/underflows can occur.
+Overflow & underflow checks are not performed for shift operations! Thus, over/underflows can occur.
 
-The left shift ``<<`` operator shifts all the beats in the first operand by the number specified in the second operand. Shifting an operand by 1 position is equivalent to multiplying it by 2, shifting 2 positions is equivalent to multiplying it by 4, and shifting 3 positions is equivalent to multiplying by 8. 
+The left shift `<<` operator shifts all the bits in the first operand by the number specified in the second operand. Shifting an operand by 1 position is equivalent to multiplying it by 2, shifting 2 positions is equivalent to multiplying it by 4 and shifting 3 positions is equivalent to multiplying by 8. 
 
 ```solidity
 uint8 public a = 100;
 uint8 public b = 2;
 
-uint8 public c = a << b; // overflow as 100 * 4 > 255
+uint8 public c = a << b;    // overflow as 100 * 4 > 255
 ```
-In the above code, left shifting ``a`` which is ``100`` by 2 positions ``b`` is equivalent to multiplying 100 by 4. So the result will overflow and the value in c will be ``144`` because ``400-256`` is ``144``.
+
+In the above code, left shifting `a` which is `100` by 2 positions `b`, is equivalent to multiplying 100 by 4. So the result will overflow and the value in c will be `144` because `400-256` is `144`.
 
 ### Use of Inline Assembly:
-Inline Assembly in solidity is performed using YUL language. In YUL programming language, integer underflow & overflow is possible as compiler does not check automatically for it as YUL is a low-level language that is mostly used for making the code more optimized, which does this by omitting many opcodes.
+
+In Solidity, inline assembly/YUL allows for low-level programming directly with EVM opcodes. This can be powerful but also risky because it lacks the built-in safety checks provided by Solidity, such as protections against overflow and underflow in arithmetic operations even if the Solidity version is >=v0.8
+
+Consider the contract below:
 
 ```solidity
 uint8 public a = 255;
 
 function addition() public returns (uint8 result) {
     assembly {
         result := add(sload(a.slot), 1) // adding 1 will overflow and reset to 0
-        // using inline assembly
     }
 
     return result;
 }
 ```
-In the above code we are adding ``1`` into the variable with inline assembly and then returning the result. The variable result will overflow and 0 will be returned, despite this the contract will not throw an error or revert.
+
+In the above code we are adding 1 into the variable with inline assembly and then returning the result. The variable result will overflow and 0 will be returned, despite this the contract will NOT throw an error or revert!
+
+
+### Subtle Overflow with Smaller Integers (e.g., `uint128`)
```

Comment:
I kinda feel like this section just leads to confusion because it's really quite similar to what would happen when not dropping down to assembly

Reply:
Well actually, my initial assumption was that it reverts when we try to return outputTokens as a uint128, but perhaps this isn't the case? If so this would be good to clarify

Reply:
> I kinda feel like this section just leads to confusion because it's really quite similar to what would happen when not dropping down to assembly

which section specifically are you talking about?

Reply:
> Well actually, my initial assumption was that it reverts when we try to return outputTokens as a uint128, but perhaps this isn't the case? If so this would be good to clarify

about this @kadenzipfel, it took me a while to wrap my head around it, but here is what I found out:-

---
1. **Function Definition:**
    ```solidity
    function getSwapQuoteUint128(uint128 amountToSwap) external view returns (uint128 outputTokens) {
        assembly {
            outputTokens := add(amountToSwap, 1)
            if lt(outputTokens, amountToSwap) { revert(0, 0) }
        }
    }
    ```
    This function is intended to return `amountToSwap + 1` and detect an overflow if `outputTokens` is less than `amountToSwap`.

2. **256-bit Arithmetic:**
    In the EVM, arithmetic operations within inline assembly (Yul) are performed using 256-bit values. When adding two 128-bit values, the result is treated as a 256-bit value.

3. **Overflow Check:**
    The function includes an overflow detection mechanism:
    ```solidity
    if lt(outputTokens, amountToSwap) { revert(0, 0) }
    ```
    This checks if `outputTokens` is less than `amountToSwap`.

4. **Overflow Scenario:**
    When `amountToSwap` is `type(uint128).max` (`2^128 - 1`), adding 1 results in `2^128`, which fits in a 256-bit register but exceeds the 128-bit range. The result is a valid 256-bit value (`2^128`), but it cannot be represented as a 128-bit value.

5. **Implicit Conversion:**
    When the 256-bit result is assigned to a 128-bit variable (`outputTokens`), it truncates the higher bits, resulting in 0 (since `2^128` overflows a 128-bit integer back to 0).

6. **Assertion:**
    The assertion confirms that the function indeed returns 0 when given the maximum `uint128` value due to the overflow:
    ```solidity
    assertEq(0, dexPair.getSwapQuoteUint128(type(uint128).max));
    ```


The key issue here is that the overflow detection mechanism:
```solidity
if lt(outputTokens, amountToSwap) { revert(0, 0) }
```
fails because `outputTokens` and `amountToSwap` are treated as 256-bit values during the addition operation. When `outputTokens` is `2^128`, it is not less than `amountToSwap` in 256-bit arithmetic. Therefore, the check `lt(outputTokens, amountToSwap)` does not trigger a revert, even though the actual result overflows when considering 128-bit constraints.

### TLDR;

The inline assembly overflow check fails because the EVM uses 256-bit arithmetic, while the function is supposed to handle 128-bit values. This discrepancy allows an overflow to evade detection, causing the function to return incorrect results without reverting. Specifically, adding 1 to the maximum `uint128` value (`2^128 - 1`) results in an overflow to 0 instead of triggering the overflow detection.

Reply:
Yeah I think this section is has too much overlap with the above section for it to be necessary